### ISSUE TITLE
[Feature] Include "Auto-sklearn metrics over time curve" as function in auto-sklearn

### ISSUE THREAD

**Issue by OliEfr** [automl/auto-sklearn#1128]
I found the "Auto-sklearn accuracy over time" plot from the  [docs](https://automl.github.io/auto-sklearn/master/examples/40_advanced/example_pandas_train_test.html#plot-the-ensemble-performance) very useful.

I would suggest to implement this natively in the auto-sklearn API. The API could either return a plot-object or the raw-data.

Would that be in your interest? LMK!

If you decide, that it would be of interest, then we need to spend some further thinking in where to implement the API. In my eyes it would be useful to have the "accuracy over time" data for regression and classification.

**Comment by mfeurer** [comment]
> Would that be in your interest? LMK!

Yes, please go ahead.

> If you decide, that it would be of interest, then we need to spend some further thinking in where to implement the API. In my eyes it would be useful to have the "accuracy over time" data for regression and classification.

I guess this function would be similar to the way `cv_results` is implemented right now. There's a main function in `automl.py` and all classes in `estimator.py` use it.

**Comment by OliEfr** [comment]
@mfeurer alright, working on it. Could you give me small hints on the following:

1) What is `testcommand.sh` for? When I try to run it, I get multiple errors about unrecognized arguments. Also, in the pytest docs I dont see any argument like `--dist`, for example.

2) I guess, that I only need to use the `Makefile` - is that right? So before "deploying", I would only run `make all`. Or do I need to do build / test someting else manually?

**Comment by mfeurer** [comment]
> What is testcommand.sh for? When I try to run it, I get multiple errors about unrecognized arguments. Also, in the pytest docs I dont see any argument like --dist, for example.

That's a script I'm using locally, not sure why that's checked in. You need a few extensions such as "pytest-xdist" and "pytest-timeout"

> I guess, that I only need to use the Makefile - is that right? So before "deploying", I would only run make all. Or do I need to do build / test someting else manually?

I'm afraid that's not up-to-date either. It suffices if you run `pytest` and `pre-commit` locally (as done on github, see the `.github` folder). If you have time we'd appreciate an update to alleviate these issues, too.

> Also, how can I build and view the docs locally? When running make docs I get doc is up to date., although I made changes (I edited the examples).

I usually go to the examples and run `make html`, not sure why it wouldn't work for you. Maybe you can delete the files generated for that example under `doc/examples` and

**Comment by OliEfr** [comment]
Hey @mfeurer,

thanks for the help. I run the pre-commit successfully. I opened a draft PR #1139. Maybe you can have a look, whether it goes in the right direction?

I am sorry, but I still dont understand how to build the docs. I think I need a bit more help here. What do you mean by 
> go to the examples

Do you mean the `doc/` folder?

I am trying to run `make html` in the `doc/` folder. I also do `sys.path.append("/home/USERNAME/miniconda3/envs/autosklearn-dev/lib/python3.8/site-packages")` in `conf.py`. However, I get the error 
> pkg_resources.DistributionNotFound: The 'scikit-learn' distribution was not found and is required by the application
when running `make html`.

**Comment by franchuterivera** [comment]
Hello @OliEfr , [this file](https://github.com/automl/auto-sklearn/blob/master/.github/workflows/docs.yml) might be a very good example of how to generate the documentation.

> Do you mean the doc/ folder?
Yes, when you run the `make html`, that also triggers the example generation that you see [here](https://automl.github.io/auto-sklearn/master/examples/index.html).

Before running `make html`, and as can be seen in the above [recommended file](https://github.com/automl/auto-sklearn/blob/master/.github/workflows/docs.yml#L18) try installing Auto-sklearn like [this](https://github.com/automl/auto-sklearn/blob/275d0d6b20d16822252d8b50bf71b1c787187f09/.github/workflows/docs.yml#L15).

Let us know if there is something else we can help with, and thanks for your contribution!

**Comment by OliEfr** [comment]
Oh god ... I was trying for hours ... apparently `pip install -e .[test]` doesn't work with ZSH-Shell .... I need to do `ip install -e ."[test]"`

**Comment by OliEfr** [comment]
@franchuterivera could you help me with a general question:

Is it somehow possible to get the output of a `print()` function in a pytest test?
I wrote some tests, however I found it difficult to debug them. I always need to run the test and return the output I am interested in order to debug - but surely, that cannot be the way to go.

I am using VSCode with Text Explorer UI and Python Text Explorer for Visual Studio Code.

I included some tests  and docs. #1139 is ready for review from my end. Some tests failed and some doc-exmpales could not be build. However, I found this didnt arise from my changes.

**Comment by franchuterivera** [comment]
> Is it somehow possible to get the output of a print() function in a pytest test?

This is a very interesting question, that depends on what you are running. The easiest answer is to use the `-s` option of pytest, something like `python -m pytest -v -s test`, which will return the stdout as you would have seen it on your console.

Nevertheless, things get complicated if you use multiprocessing, because a worker will run on a different process. But after looking at your code, just using `-s` should do the trick.

One recommendation! probable you want to run the example and fit a classifier with the name estimator, and at the end do something like:
```
    with open('estimator.pickle', 'wb') as handle:
        pickle.dump(api, handle, protocol=pickle.HIGHEST_PROTOCOL)

    # Then let us read it back and use it for our analysis
    with open('estimator.pickle', 'rb') as handle:
        estimator = pickle.load(handle)

```
In case you easily need to debug, you can just load the estimator instead of waiting for `fit()` to complete.

**Comment by mfeurer** [comment]
Hey, is the corresponding PR merged yet?

**Comment by eddiebergman** [comment]
Yes it's been merged, I'll close the issue!
